src/components/TradeWidget/TokenRow.tsx

```diff
@@ -144,12 +204,26 @@ const TokenRow: React.FC<Props> = ({ token, tokens, selectLabel, onSelectChange,
         />
       </SelectBox>
       <InputBox>
-        <input type="text" placeholder="0" />
+        <input
+          className={className}
+          placeholder="0"
+          name={inputId}
+          type="text"
+          required
+          ref={register({
+            validate: {
+              positive: (value: string): true | string => validatePositive(value),
+            },
+          })}
+        />
+        {errorOrWarning}
         <WalletDetail>
           <strong>
             <Link to="/deposit">Exchange wallet:</Link>
           </strong>{' '}
-          <span className="success">{displayBalance(balance, 'exchangeBalance')}</span>
+          <a className="success" onClick={(): void => setValue(inputId, max.toFixed(4))}>
```

Comment:
`toFixed` rounds the number, `max.toFixed(4)` can be `> max`, e.g. `5.55555.toFixed(4) = '5.5556'` so it'll error by default

Reply:
which is why I find max dangerous in the first place tbh